fix: Major refactoring of Polling, Retry and Timeout logic

vam-google opened the pull request:
If you are a  user of python-api-core and experience polling timing out after ~15 minutes (900s) in your python code after this change, please make sure that instead of calling `PollingFuture.result()` you call it with additonal `timeout` argument like this: `PollingFuture.result(timeout = <desired timeout in seconds>)` or `PollingFuture.result(timeout = None)` (for infinite timeout, but infinite timeouts are strongly **discouraged**) for  your code to stop timing out at 900s.

The core libraries cannot and should not have infitine polling as as default behavior. The fact that it had it like that in python for years (unlike other GCP-supported languages) was a bug, as it contradicts original cross-language LRO methods design in GAPIC libraries and in general users are not supposed to be put into an infinite-loop scenario implicitly. If an LRO can actually ran for hours or even days and it is Ok, then users are expected to acknowledge this fact by providing the huge timeouts explicitly.

This is in response to https://freeman.vc/notes/aws-vs-gcp-reliability-is-wildly-different, which triggered an investigation of the whole Polling/Retry/Timeout behavior in Python GAPIC clients and revealed many fundamental flaws in its implementation.

To properly describe the refactoring in this PR we need to stick to a rigorous terminology, as vague definitions of retries, timeouts, polling and related concepts seems to be the main source of the present bugs and overall confusion among both groups: users of the library and creators of the library. Please check the updated (in this PR) documentation of the `google.api_core.retry.Retry` class and the `google.api_core.future.polling.Polling.result()` method for the proper definitions and context.

Note, the overall semantics around Polling, Retry and Timeout remains quite confusing even after refactoring (although it is now more or less rigorously defined), but it was as clean as I could make it while still maintaining backward compatibility of the whole library.

The quick summary of the changes in this PR:

1) Properly define and fix the application of Deadline and Timeout concepts. Please check the updated documentation for the `google.api_core.retry.Retry` class for the actual definitions. Originally the `deadline` has been used to represent timeouts conflating the two concepts. As result this PR replaces `deadline` arguments with `timeout` ones in as backward-compatible manner as possible (i.e. backward compatible in all practical applications).

2) Properly define RPC Timeout, Retry Timeout and Polling Timeout and how a generic Timeout concept (aka Logical Timeout) is mapped to one of those depending on the context. Please check `google.api_core.retry.Retry` class documentation for details.

3) Properly define and fix the application of Retry and Polling concepts. Please check the updated documentation for `google.api_core.future.polling.PollingFuture.result()` for details.

4) Separate `retry` and `polling` configurations for Polling future, as these are two different concepts (although both operating on `Retry` class). Originally both retry and polling configurations were controlled by a single `retry` parameter, merging configuration regarding how "rpc error responses" and how "operation not completed" responses are supposed to be handled.

5) For the following config properties - `Retry` (including `Retry Timeout`), `Polling` (including `Polling Timeout`) and `RPC Timeout` - fix and properly define how each of the above properties gets configured and which config gets precedence in case of a conflict (check `PollingFuture.result()` method documentation for details). Each of those properties can be specified as follows: directly provided by the user for each call, specified during gapic generation time from config values in `grpc_service_config.json` file (for Retry and RPC Timeout) and `gapic.yaml` file (for Polling Timeout), or be provided as a hard-coded basic default values in python-api-core library itself. This alo includes fixing the per-call polling config propagation logic (the polling/retry configs supplied to `PollingFuture.result()` used to be ignored for actual call).

6) Deprecate `ExponentialTimeout`, `ConstantTimeout` and related logic as those are outdated concepts and are not consistent with the other GAPIC Languages. Replace it with `TimeToDeadlineTimeout` to be consistent with how the rest of the languages do it.

7) Deprecate `google.api_core.operations_v1.config` as it is an outdated concept and self-inconsistent (as all gapic clients provide configuraiton in code). The configs are directly provided in code instead.

8) Switch randomized delay calculation from `delay` being treated as expected value for randomized_delay to `delay` being treated as maximum value for `randomized_delay` (i.e. the new expected valud for `randomized_delay` is `delay / 2`). See the `exponential_sleep_generator()` method implementation for details. This is needed to make Python implementation of retries and polling exponential backoff consistent with the rest of GAPIC languages. Also fix the uncontrollable growth of `delay` value (since it is a subject of exponential growth, the `delay` value was quickly reaching "infinity" value, and the whole thing was not failing simply due to python being a very forgiving language which forgives multiplying "infinity" by a number (`inf * number = inf`) binstead of simply overflowing to a (most likely) negative number). Also essentially rollback  the https://github.com/googleapis/python-api-core/commit/52f12af027568ddeae7fb6b1e3d2111cce1bac08 change, since that is inconsistent with the other languages and damages uniform distibution of retry delays artificially shifting  their concentration towards the end of timeout.

9) Fix url construction in `OperationsRestTransport`. Without this fix the polling logic for REST transport was completely broken (is not affecting Compute client, as that one has custom LRO).

10) Last but not least: change the default values for Polling logic to be the following: `initial=1.0` (same as before), `maximum=20.0` (was `60`), `multiplier=1.5` (was `2.0`), `timeout=900` (was `120`, but due to timeout resolution logic was actually None (i.e. infinity)). This, in conjunction with changed calculation of randomized delay (i.e. its expected value now being  `delay / 2`) overall makes polling logic much less aggressive in terms of increasing delays between each polling iteration, making LRO return much earlier for users on average, but still keeping a healthy balance between strain put on both client and server by polling and responsiveness of LROs for user.

*The design doc summarising all the changes and reasons for them is in progress.

In addition to the timeout/retry fixes, this PR has some other non-related technical fixes: 
- Fix and improve code coverage (and explicitly disables lines which are not supposed to be covered) 
- Define `Python 3.10` as the default python version used in CI.
- Upgrade to Sphinx 4.2.0, as it is the one supporting Python 3.10

vam-google committed:
fix: Major refactoring and fix for Polling, Retry and Timeout logic

This is in response to https://freeman.vc/notes/aws-vs-gcp-reliability-is-wildly-different, which triggered an investigation of the whole Polling/Retry/Timeout behavior in Python GAPIC clients and revealed many fundamental flaws in its implementaiton.

To properly describe the refactoring this PR does we need to stick to a rigorous terminology, as vague definitions of retries, timeouts, polling and related concepts seems to be the main source of the present bugs and overal confusion among both groups: users of the library and creators of the library. Please check the documentation of the `google.api_core.retry.Retry` class the `google.api_core.future.polling.Polling.result()` method for the proper definitions and context.

Note, the overall semantics around Polling, Retry and Timeout remains quite confusing even after refactoring (although it is now more or less rigorously defined), but it was clean as I could make it while still maintaining backward compatibility of the whole library.

The quick summary of the changes in this PR:

1) Properly define and fix the application of Deadline and Timeout concepts. Please check the updated documentation for the `google.api_core.retry.Retry` class for the actual definitions. Originally the `deadline` has been used to represent timeouts conflating the two concepts. As result this PR replaces `deadline` arguments with `timeout` ones in as backward-compatible manner as possible (i.e. backward compatible in all practical applications).

2) Properly define RPC Timeout, Retry Timeout and Pollint Timeout and how a generic Timeout concept (aka Logical Timeout) is mapped to one of those depending on the context. Please check `google.api_core.retry.Retry` class documentation for details.

3) Properly define and fix the application of Retry and Polling concepts. Please check the updated documentation for `google.api_core.future.polling.PollingFuture.result()` for details.

4) Separate `retry` and `polling` configurations for Polling future, as these are two different concepts (although both operating on `Retry` class). Originally both retry and polling configurations were controlled by a single `retry` parameter, merging configuration regarding how "rpc error responses" and how "operation not completed" responses are supposed to be handled.

5) For the following config properties - `Retry (including `Retry Timeout`), `Polling` (including `Polling Timeout`) and `RPC Timeout` - fix and properly define how each of the above properties gets configured and which config gets precedence in case of a conflict (check `PollingFuture.result()` method documentation for details). Each of those properties can be specified as follows: directly provided by the user for each call, specified during gapic generation time from config values in `grpc_service_config.json` file (for Retry and RPC Timeout) and `gapic.yaml` file (for Polling), or be provided as a hard-coded basic default values in python-api-core library itself.

6) Fix the per-call polling config propagation logic (the polling/retry configs supplied to `PollingFuture.result()` used to be ignored for actual call).

7) Deprecate the usage of `deadline` terminology in the whole library and backward-compatibly replace it with timeout. This is essential as what has been called "deadline" in this library was actually "timeout" as it is defined in `google.api_core.retry.Retry` class documentation.

8) Deprecate `ExponentialTimeout`, `ConstantTimeout` and related logic as those are outdated concepts and are not consistent with the other GAPIC Languages. Replace it with `TimeToDeadlineTimeout` to be consistent with how the rest of the languages do it.

9) Deprecate `google.api_core.operations_v1.config` as it is an outdated concept and self-inconsistent (as all gapic clients provide configuraiton in code). The configs are directly provided in code instead.

10) Switch randomized delay calculation from `delay` being treated as expected value for randomized_delay to `delay` being treated as maximum value for `randomized_delay` (i.e. the new expected valud for `randomized_delay` is `delay / 2`). See the `exponential_sleep_generator()` method implementation for details. This is needed to make Python implementation of retries and polling exponential backoff consistent with the rest of GAPIC languages. Also fix the uncontrollable growth of `delay` value (since it is a subject of exponential growth, the `delay` value was quickly reaching "infinity" value, and the whole thing was not failing simply due to python being a very forgiving language which forgives multiplying "infinity" by a number (`inf * number = inf`) binstead of simply overflowing to a (most likely) negative number).

11) Fix url construction in `OperationsRestTransport`. Without this fix the polling logic for REST transport was completely broken (is not affecting Compute client, as that one has custom LRO).

12) Las but not least: change the default values for Polling logic to be the following: `initial=1.0` (same as before), `maximum=20.0` (was `60`), `multiplier=1.5` (was `2.0`), `timeout=900` (was `120`, but due to timeout resolution logic was actually None (i.e. infinity)). This, in conjunction with changed calculation of randomized delay (i.e. its expected value now being  `delay / 2`) overall makes polling logic much less aggressive in terms of increasing delays between each polling iteration, making LRO return much earlier for users on average, but still keeping a healthy balance between strain put on both client and server by polling and responsiveness of LROs for user.

*The design doc summarising all the changes and reasons for them is in progress.

vam-google committed:
fix ci failures (mainly sphinx errors)

atulep commented on google/api_core/future/polling.py:
```diff
@@ -29,14 +29,40 @@ class _OperationNotComplete(Exception):
+# (to not be confused with polling logic).
+DEFAULT_RETRY = retries.Retry(predicate=RETRY_PREDICATE)
+
+# Poling predicate is supposed to poll only on _OperationNotComplete.
```
Polling

vam-google committed:
remove unused code

vam-google committed:
fix typo

vam-google committed:
Pin pytest version to <7.2.0

vam-google committed:
reformat code

atulep commented on google/api_core/extended_operation.py:
```diff
@@ -68,9 +71,14 @@ class ExtendedOperation(polling.PollingFuture):
     """
 
     def __init__(
-        self, extended_operation, refresh, cancel, retry=polling.DEFAULT_RETRY
```
Right now, setting retry will be no-op. It will be a confusing user behavior. Ideally, we would warn developer somehow. I don't think we handle this in GAPIC right now. Something to consider - https://docs.python.org/3/library/exceptions.html#DeprecationWarning.

vam-google replied:
Setting retry will not be no-op, it will overwrite settng polling for backward compatibility int `PollingFuture.__init__()`:
```
self._polling = kwargs.get("retry", polling)
```

The deprecation warning is not thrown to avoid contaminating users logs, since it will be affecting every current user.

atulep commented on google/api_core/extended_operation.py:
```diff
@@ -114,7 +122,7 @@ def error_message(self):
 
-    def done(self, retry=polling.DEFAULT_RETRY):
+    def done(self, retry=None):
         self._refresh_and_update(retry)
```
What happens if user wants to use a different polling other than DEFAULT_RETRY ?

atulep commented on google/api_core/extended_operation.py:
```diff
@@ -137,9 +145,11 @@ def cancelled(self):
         if not self._extended_operation.done:
-            self._extended_operation = self._refresh(retry=retry)
+            self._extended_operation = (
+                self._refresh(retry=retry) if retry else self._refresh()
```
Simplify with self._refresh(retry=retry)

vam-google replied:
if `retry=None` then `self._refresh(retry=retry)` and `self._refresh()` are different things, becuase `_refresh` is  most likely defined as `_refrhesh(retry=DEFAULT)`. Note in the whole semantics of gapics `retry/polling` configs `None` has special meaning, and is different from being "not set" or "default".

atulep commented on google/api_core/future/polling.py:
```diff
@@ -29,14 +29,37 @@ class _OperationNotComplete(Exception):
+    initial=1.0,
+    maximum=20.0,
+    multiplier=1.5,
+    timeout=900,
```
Add comment with units.

vam-google replied:
added

atulep commented on google/api_core/future/polling.py:
```diff
@@ -69,57 +100,149 @@ def __init__(self, retry=DEFAULT_RETRY):
             return
 
-        retry_ = self._retry.with_deadline(timeout)
+        polling = polling or self._polling
```
I wouldn't do it. It's confusing if I don't pass a `polling` or explcitly set it to None, but function will overwrite it. Why don't remove from arg list?

vam-google replied:
Because this the contract all gapic libraries have. Plling or retry configs can be set by users explicitly or, if not set, default ones will be applied. This is a cross-langauge concept. As for why polling was even added in the args in this PR: this is to fix bug in a backward compatible way: originally `retry` argument was viewed as the one controlling polling, although it does not do it, so a new polling argument was added to provide that functionality to the users (i.e. to have an ability to config polling per method call).

atulep replied:
> so a new polling argument was added to provide that functionality to the users
Users can already configure polling in the constructor. This functionality already exists right ?

vam-google replied:
In which constructor? Please be more specific. Do you mean PollingFuture constructor? Then yes. This is how it works: PollingFuture has a polling config (`self._polling`) which will be  used if user did not specify polling explicitly when they call `future.result()` (i.e. use default one, when "default" is the one passed in the constructor, if the user did not specify their custom config).

atulep replied:
Okay, this is an anti-pattern then. If you want to configure per method, it shouldn't be in the constructor. However, I see your point that it will be backwards incompatible to remove from constructor, and need to keep per-method config option. No further action needed.

atulep commented on google/api_core/future/polling.py:
```diff
@@ -69,57 +100,149 @@ def __init__(self, retry=DEFAULT_RETRY):
         except exceptions.RetryError:
             raise concurrent.futures.TimeoutError(
-                "Operation did not complete within the designated " "timeout."
+                "Operation did not complete within the designated timeout."
```
Can you include timeout value in the exception ?

vam-google replied:
added

vam-google committed:
address pr feedback

vam-google commented:
@atulep Addressed your comments PTAL

atulep commented on tests/asyncio/gapic/test_method_async.py:
```diff
@@ -198,41 +198,6 @@ async def test_wrap_method_with_overriding_retry_and_timeout(unused_sleep):
-    autospec=True,
-)
-@pytest.mark.asyncio
-async def test_wrap_method_with_overriding_retry_deadline(utcnow, unused_sleep):
```
Why was this removed ?

vam-google replied:
Just updated the PR description. The last few sentences of the item 8 describe the removal of unnecessary logic, and this test was testing it. From the PR description for your convenience

> Also essentially rollback the https://github.com/googleapis/python-api-core/commit/52f12af027568ddeae7fb6b1e3d2111cce1bac08 change, since that is inconsistent with the other languages and damages uniform distibution of retry delays artificially shifting their concentration towards the end of timeout.

atulep commented on tests/asyncio/test_retry_async.py:
```diff
@@ -277,7 +277,7 @@ async def test___call___and_execute_success(self, sleep):
         sleep.assert_not_called()
 
     # Make uniform return half of its maximum, which is the calculated sleep time.
-    @mock.patch("random.uniform", autospec=True, side_effect=lambda m, n: n / 2.0)
```
Why was this removed ?

vam-google replied:
This is also a result of the change described in the item 8 of the PR description:

> Switch randomized delay calculation from delay being treated as expected value for randomized_delay to delay being treated as maximum value for randomized_delay (i.e. the new expected valud for randomized_delay is delay / 2). See the exponential_sleep_generator() method implementation for details. This is needed to make Python implementation of retries and polling exponential backoff consistent with the rest of GAPIC languages.

atulep commented on tests/unit/test_retry.py:
```diff
@@ -406,8 +407,17 @@ def increase_time(sleep_delay):
 
-        assert last_wait == 2.9  # and not 8.0, because the last delay was shortened
-        assert total_wait == 9.9  # the same as the deadline
+        assert last_wait == 8.0
```
Why is it 8? i actually don't get

vam-google replied:
Same thing as the other two - all are the effect of the removal of the rounding for the last attempt.

atulep commented on tests/unit/gapic/test_method.py:
```diff
@@ -199,35 +178,35 @@ def test_wrap_method_with_overriding_retry_and_timeout(unusued_sleep):
-        + 5  # First to set the deadline.
-        + 5  # One for each min(timeout, maximum, (DEADLINE - NOW).seconds)
-    )
+# @mock.patch("time.sleep")
```
Remove

vam-google replied:
removed

vam-google committed:
address PR feedback

vam-google commented:
@atulep Addressed PR feedback, PTAL. Also let me know if you have any more questions (tl;dr; the changes you were commenting on in your last review cycle were all affected by removal of this logic: https://github.com/googleapis/python-api-core/commit/52f12af027568ddeae7fb6b1e3d2111cce1bac08#diff-c118188be6311975226123f406396979b3dbb200e36b425824fba2aa2d675fa7R208, which should not have been there in the first place.

atulep commented on google/api_core/retry.py:
```diff
@@ -139,15 +139,15 @@ def exponential_sleep_generator(initial, maximum, multiplier=_DEFAULT_DELAY_MULT
         float: successive sleep intervals.
     """
-    delay = initial
+    delay = min(initial, maximum)
```
Are they never negative ?

vam-google replied:
Who are they? initial and maximum? Yes, they are. There is still a chance that a user will set the values as negative, but it is very unlikely and could be done only either by mistake or with intention to break it all. We could have validated that the values are not negative but seems like gapic libraries do not perform these kinds of validation (i.e. by convention we don't go as far as validating ranges of all integer params), so not validating it here (this is not changed in this PR and has been like this before)

atulep commented on google/api_core/future/polling.py:
```diff
@@ -69,57 +100,150 @@ def __init__(self, retry=DEFAULT_RETRY):
             return
 
-        retry_ = self._retry.with_deadline(timeout)
+        polling = polling or self._polling
```
Helpful to add a comment of what type Polling is. Doesn't need to be a formal Pytype type, but something to help the reader.

vam-google replied:
What do you mean by "type polling"? Which types of polling are there?

atulep replied:
Python type of the `polling` variable.

vam-google replied:
I don't understand what you mean. Every polling config is of type Retry.

atulep commented on google/api_core/gapic_v1/method.py:
```diff
@@ -52,65 +52,26 @@ def _apply_decorators(func, decorators):
     return func
 
 
-def _determine_timeout(default_timeout, specified_timeout, retry):
```
Correct me here - this is removed because timeout is being passed as an argument to _blocking_poll ?

vam-google replied:
This is because ExponentialTimeout and ConstantTimeout are both deprecated and should not be used. This method was essentially creating/selecting one of the two deprecated timeout implementations, since both of the timeouts are not needed and the method is prefixed with `_` (not supposed to be used publicly) i just removed it.

atulep commented on tests/unit/test_retry.py:
```diff
@@ -361,7 +362,7 @@ def test___call___and_execute_retry(self, sleep, uniform):
         assert on_error.call_count == 1
 
     # Make uniform return half of its maximum, which is the calculated sleep time.
```
Update the comment here to remove halfing.

vam-google replied:
Good catch! removed

atulep commented:
Looks good overall. Have last few comments.

vam-google committed:
address pr feedback

alexander-fenster commented:
@vam-google notification test

vam-google commented:
@vam-google test

vchudnov-g commented on google/api_core/retry.py:
```diff
@@ -223,22 +224,88 @@ def retry_target(target, predicate, sleep_generator, deadline, on_error=None):
+    of this class, and it defines how long a certain RPC may be retried in case
+    an error is returned from the server.
+
+    Polling Timeout - it is similar to Retry Timeout, but defines how long the
+    client side is allowed to call polling rpc repeatedly to check a status of a
+    long running operaiton. Unlike in the retry case, the polling rpc is
+    expected to succed (its errors are supposed to be handled by the retry
+    logic). The decision if a new polling attemtp needs to be made is made
+    not based on the RPC status code but based on the status of the returned
+    status of an operation (i.e. it is higher level concept than the rpc error
+    codes).
```
```suggestion
    Polling Timeout: defines how long the
    client side is allowed to call the polling RPC repeatedly to check a status of a
    long-running operation. Each polling RPC is
    expected to succeed (its errors are supposed to be handled by the retry
    logic). The decision as to whether a new polling attempt needs to be made is based
    not on the RPC status code but  on the status of the returned
    status of an operation. In other words: we will poll a long-running operation until the operation is done or the polling timeout expires. Each poll will inform us of the status of the operation. The poll consists of an RPC to the server that may itself be retried as per the poll-specific retry settings in case of errors. The operation-level retry settings do NOT apply to polling-RPC retries.
```

I tried to make this description a bit clearer. I think it's better but still needs some word-smithing.

vchudnov-g commented on google/api_core/retry.py:
```diff
@@ -223,22 +224,88 @@ def retry_target(target, predicate, sleep_generator, deadline, on_error=None):
+    that is. In that case the actual timeout type (sometimes also refered to as
+    Logical Timeout) can be determined from the context. If it is a unary rpc
+    call (i.e. a regular one) Timeout usually stands for the RPC Timeout (if
+    provided directly as a standaone value) or Retry Timeout (if provided as
```
```suggestion
    provided directly as a standalone value) or Retry Timeout (if provided as
```

vchudnov-g commented on google/api_core/retry.py:
```diff
@@ -223,22 +224,88 @@ def retry_target(target, predicate, sleep_generator, deadline, on_error=None):
+    call (i.e. a regular one) Timeout usually stands for the RPC Timeout (if
+    provided directly as a standaone value) or Retry Timeout (if provided as
+    ``retry.timeout`` property of the unary RPC's retry config). For
+    ``Operation`` or ``PolllingFuture`` in general Timeout stands for
```
```suggestion
    ``Operation`` or ``PollingFuture`` in general Timeout stands for
```

vchudnov-g commented on google/api_core/retry.py:
```diff
@@ -223,22 +224,88 @@ def retry_target(target, predicate, sleep_generator, deadline, on_error=None):
-        deadline (float): How long to keep retrying in seconds. The last sleep
-            period is shortened as necessary, so that the last retry runs at
-            ``deadline`` (and not considerably beyond it).
+        timeout (float): How long to keep retrying in seconds.
```
```suggestion
        timeout (float): How long to keep retrying, in seconds.
```

vchudnov-g commented on google/api_core/retry.py:
```diff
@@ -223,22 +224,88 @@ def retry_target(target, predicate, sleep_generator, deadline, on_error=None):
             if the given exception is retryable.
         initial (float): The minimum amount of time to delay in seconds. This
             must be greater than 0.
         maximum (float): The maximum amount of time to delay in seconds.
```
```suggestion
        maximum (float): The maximum amount of time to delay, in seconds.
```

vchudnov-g commented on google/api_core/retry.py:
```diff
@@ -223,22 +224,88 @@ def retry_target(target, predicate, sleep_generator, deadline, on_error=None):
-            ``deadline`` (and not considerably beyond it).
+        timeout (float): How long to keep retrying in seconds.
+        deadline (float): DEPRECATED: use `timeout` instead. For backward
+            compatibility, if specified it will override ``timeout`` parameter.
```
```suggestion
            compatibility, if specified it will override the ``timeout`` parameter.
```

vchudnov-g commented on google/api_core/retry.py:
```diff
@@ -284,31 +353,53 @@ def retry_wrapped_func(*args, **kwargs):
     def deadline(self):
-        return self._deadline
+        """
+        DEPRECATED: use ``timeout`` instead.  Check the ``Retry`` class
```
```suggestion
        DEPRECATED: use ``timeout`` instead.  Refer to the ``Retry`` class
```

vchudnov-g commented on google/api_core/retry.py:
```diff
@@ -284,31 +353,53 @@ def retry_wrapped_func(*args, **kwargs):
-        """Return a copy of this retry with the given deadline.
+        """Return a copy of this retry with the given timeout.
+
+        DEPRECATED use :meth:`with_timeout` instead. Check the ``Retry`` class
```
```suggestion
        DEPRECATED: use :meth:`with_timeout` instead. Refer to the ``Retry`` class
```

vchudnov-g commented on google/api_core/retry.py:
```diff
@@ -284,31 +353,53 @@ def retry_wrapped_func(*args, **kwargs):
 
         Args:
-            deadline (float): How long to keep retrying.
+            timeout (float): How long to keep retrying in seconds.
```
```suggestion
            timeout (float): How long to keep retrying, in seconds.
```

vchudnov-g commented on google/api_core/retry_async.py:
```diff
@@ -68,9 +68,12 @@ async def check_if_exists():
 
 
-async def retry_target(target, predicate, sleep_generator, deadline, on_error=None):
+async def retry_target(
```
nit: any way to make the formatting nicer?

vam-google replied:
Please do not suggest formatting, it is autoformatted

vchudnov-g commented on google/api_core/retry_async.py:
```diff
@@ -84,12 +87,12 @@ async def retry_target(target, predicate, sleep_generator, deadline, on_error=No
-        deadline (float): How long to keep retrying the target. The last sleep
-            period is shortened as necessary, so that the last retry runs at
-            ``deadline`` (and not considerably beyond it).
+        timeout (float): How long to keep retrying the target.
```
```suggestion
        timeout (float): How long to keep retrying the target, in seconds.
```

vchudnov-g commented on google/api_core/retry_async.py:
```diff
@@ -84,12 +87,12 @@ async def retry_target(target, predicate, sleep_generator, deadline, on_error=No
         on_error (Callable[Exception]): A function to call while processing a
             retryable exception.  Any error raised by this function will *not*
             be caught.
+        deadline (float): DEPRECATED use ``timeout`` instead. For beckward
+        compatibility, if set it will override ``timeout`` parameter.
```
```suggestion
        deadline (float): DEPRECATED use ``timeout`` instead. For backward
        compatibility, if set it will override the ``timeout`` parameter.
```

vchudnov-g commented on google/api_core/retry_async.py:
```diff
@@ -228,28 +236,40 @@ def _replace(
+        """Return a copy of this retry with the given timeout.
+
+        Args:
+            timeout (float): How long to keep retrying.
```
```suggestion
            timeout (float): How long to keep retrying, in seconds.
```

vchudnov-g commented on google/api_core/retry_async.py:
```diff
@@ -228,28 +236,40 @@ def _replace(
 
     def with_deadline(self, deadline):
         """Return a copy of this retry with the given deadline.
+        DEPRECATED use :meth:`with_timeout` instead.
```
```suggestion
        DEPRECATED: use :meth:`with_timeout` instead.
```

vchudnov-g commented on google/api_core/timeout.py:
```diff
@@ -15,7 +15,8 @@
 """Decorators for applying timeout arguments to functions.
 
 These decorators are used to wrap API methods to apply either a constant
-or exponential timeout argument.
+(DEPRECATED), exponential (DEPRECATED) or Deadline-dependent (recommended)
```
put the recommended one first:
"to apply a Deadline-dependent (RECOMMENDED), constant (DEPRECATED), or exponential (DEPRECATED) "

vam-google replied:
Flipped the order

vchudnov-g commented on google/api_core/timeout.py:
```diff
@@ -66,9 +67,69 @@ def is_thing_ready(timeout=None):
 class ConstantTimeout(object):
     """A decorator that adds a constant timeout argument.
 
+    DEPRECATED: use `TimeToDeadlineTimeout` instead.
```
```suggestion
    DEPRECATED: use ``TimeToDeadlineTimeout`` instead.
```

vchudnov-g commented on google/api_core/timeout.py:
```diff
@@ -140,6 +201,9 @@ def _exponential_timeout_generator(initial, maximum, multiplier, deadline):
     """A decorator that adds an exponentially increasing timeout argument.
 
+    DEPRECATED: the concept of incrementing timeout exponentially has been
+    deprecated. Use `TimeToDeadlineTimeout` instead.
```
```suggestion
    deprecated. Use ``TimeToDeadlineTimeout`` instead.
```

vchudnov-g commented on tests/unit/test_general_helpers.py:
```diff
@@ -0,0 +1,13 @@
+# Copyright 2022, Google LLC All rights reserved.
```
Let's delete this file if it's not doing anything.

vam-google replied:
It is necesssary for test coverage to do its thing. It is an empty test for an empty file (I don't know the history of that, just trying to fix a very broken build here).

vchudnov-g reviewed (commented):
I have not looked carefully at the tests yet, but so far this looks great. I will be running some manual tests to complement what Tony and Aza have been doing.

Most of the comments are documentation tweaks, so hopefully you can just accept them on GitHub and pull them into your local repo before making any other changes.

vam-google committed:
Update google/api_core/future/polling.py

Co-authored-by: Victor Chudnovsky <vchudnov@google.com>

vam-google committed:
Apply documentation suggestions from code review

Co-authored-by: Victor Chudnovsky <vchudnov@google.com>

vam-google commented on google/api_core/retry.py:
```diff
@@ -198,19 +201,17 @@ def retry_target(target, predicate, sleep_generator, deadline, on_error=None):
+            )
+            if deadline < next_attempt_time:
                 raise exceptions.RetryError(
                     "Deadline of {:.1f}s exceeded while calling target function".format(
-                        deadline
+                        timeout
                     ),
```
This is all automatically formatted, please do not do formatting suggestions as they would violate formatting validation anyways.

vam-google commented on google/api_core/retry.py:
```diff
@@ -198,19 +201,17 @@ def retry_target(target, predicate, sleep_generator, deadline, on_error=None):
+            )
+            if deadline < next_attempt_time:
                 raise exceptions.RetryError(
                     "Deadline of {:.1f}s exceeded while calling target function".format(
```
No, it is one of the places where it is actually deadline.

vam-google commented on google/api_core/future/polling.py:
```diff
@@ -29,14 +29,37 @@ class _OperationNotComplete(Exception):
 )
-DEFAULT_RETRY = retry.Retry(predicate=RETRY_PREDICATE)
+
+# DEPRECATED, use DEFAULT_POLLING to configure LRO polling logic. Construct
```
The baseliine part was there before. Just keeping it as is, as the current recommendation is simply to not used that at all for anything.

vam-google replied:
As for being explicit on usage of DEFAULT_RETRY, it is deprecated, so it should not be used.

vam-google commented on google/api_core/future/polling.py:
```diff
@@ -69,57 +100,150 @@ def __init__(self, retry=DEFAULT_RETRY):
 
-    def _blocking_poll(self, timeout=None, retry=DEFAULT_RETRY):
-        """Poll and wait for the Future to be resolved.
+    def _blocking_poll(self, timeout=_DEFAULT_VALUE, retry=None, polling=None):
```
Here it is not a refactoring it is a fix, so it does not intend to be backward-compatible (as old behavior was wrong).
It used to be:
1) retry ignored
2) timeout always None, making actual timeout from config also being ignored.

That is being fixed here:
1) Retry is now appreciated as retry config for status checking rpc
2) Timeout will be None (meaning infinite) only if explicitly specified, otherwise it will be the defatult one or from config file (how that is resovled is described in the internal design doc)

vam-google commented on google/api_core/future/polling.py:
```diff
@@ -69,57 +100,150 @@ def __init__(self, retry=DEFAULT_RETRY):
-                How long (in seconds) to wait for the operation to complete.
-                If None, wait indefinitely.
+            timeout (int): (Optional) How long (in seconds) to wait for the
+                operation to complete.
```
Added note about None. As for the rest of behavior it is described int the bigger comment for this method. Note, resolution of actual config values used in runtime is surprisingly complex, so it is impossible to describe that in a comment here (there is a whole section in retry design doc describing it).

vam-google commented on google/api_core/future/polling.py:
```diff
@@ -128,8 +252,8 @@ def result(self, timeout=None, retry=DEFAULT_RETRY):
-        kwargs = {} if retry is DEFAULT_RETRY else {"retry": retry}
-        self._blocking_poll(timeout=timeout, **kwargs)
+
+        self._blocking_poll(timeout=timeout, retry=retry, polling=polling)
```
It is not backward compatible by design, because it is a fix (this whole PR is not a refactoring only it also contains many fixes, this is one of them.

vam-google commented on google/api_core/gapic_v1/method.py:
```diff
@@ -52,65 +52,26 @@ def _apply_decorators(func, decorators):
+            callable (i.e. duration of time within which an RPC must terminate
+            after its start, to not be confused with deadline). If ``None``,
+            this callable will not specify a timeout argument to the low-level
+            RPC method by default.
```
The "by default" was just redundant, removed it.

vam-google commented on google/api_core/operations_v1/transports/rest.py:
```diff
@@ -73,6 +74,7 @@ def __init__(
         always_use_jwt_access: Optional[bool] = False,
         url_scheme: str = "https",
         http_options: Optional[Dict] = None,
+        uri_prefix: str = "v1",
```
Ok, thought it will require a change in gapic-generator-python PR as well

vam-google commented on google/api_core/retry.py:
```diff
@@ -161,12 +161,14 @@ def retry_target(target, predicate, sleep_generator, deadline, on_error=None):
         sleep_generator (Iterable[float]): An infinite iterator that determines
             how long to sleep between retries.
-        deadline (float): How long to keep retrying the target. The last sleep
+        timeout (float): How long to keep retrying the target. The last sleep
```
good catch. The whole "shortening" logic has been removed, so I removed the part of the comment talking about it (including the reference to deadline)

vam-google committed:
Address PR feedback

tswast commented on google/api_core/future/polling.py:
```diff
@@ -69,57 +100,150 @@ def __init__(self, retry=DEFAULT_RETRY):
+            polling = polling.with_timeout(timeout)
 
         try:
-            kwargs = {} if retry is DEFAULT_RETRY else {"retry": retry}
```
FYI: this was a breaking change. https://github.com/googleapis/python-api-core/issues/477

sasha-gitg replied:
Also for python-aiplatform: https://github.com/googleapis/python-aiplatform/issues/1870

vam-google replied:
that rety logic line never worked (the retry had been i. What broke you is most likely the new default timeout value (instead of None).

jacobg commented:
It seems like this merge broke using freezegun, particularly with google-cloud-datastore in unit testing. Here's the stack trace:
```
  File "/Users/jacob/.pyenv/versions/gae2/lib/python3.9/site-packages/google/cloud/datastore/transaction.py", line 231, in begin
    response_pb = self._client._datastore_api.begin_transaction(
  File "/Users/jacob/.pyenv/versions/gae2/lib/python3.9/site-packages/google/cloud/datastore_v1/services/datastore/client.py", line 834, in begin_transaction
    response = rpc(
  File "/Users/jacob/.pyenv/versions/gae2/lib/python3.9/site-packages/google/api_core/gapic_v1/method.py", line 113, in __call__
    return wrapped_func(*args, **kwargs)
  File "/Users/jacob/.pyenv/versions/gae2/lib/python3.9/site-packages/google/api_core/timeout.py", line 120, in func_with_timeout
    return func(*args, **kwargs)
  File "/Users/jacob/.pyenv/versions/gae2/lib/python3.9/site-packages/google/api_core/grpc_helpers.py", line 74, in error_remapped_callable
    raise exceptions.from_grpc_error(exc) from exc
google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded
```

Would you expect that to happen? Any workarounds?
